File: depth-first-search/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        return list(dfs(root1, [])) == list(dfs(root2, []))
    
def dfs(root, array):
    if root is None:
        return
    if root.left is None and root.right is None:
        array.append(root.val)
    dfs(root.left, array)
    dfs(root.right, array)
    return array

        

File: tree/leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    const arr1 = []
    const arr2 = []

    const dfs = (root, arr) => {
        if (!root.left && !root.right) {
            arr.push(root.val)
        }
        dfs(root.left, arr)
        dfs(root.right,arr)
    }
    dfs(root1,arr1)
    dfs(root2,arr2)
    if (arr1.length !== arr2.length) return false;

    for (let i = 0; i < arr1.length; i++) {
        if (arr1[i] !== arr2[i]) return false;
    }

    return true;    
};